M2/Macaulay2/packages/IncidenceCorrespondenceCohomology.m2

```diff
@@ -0,0 +1,2105 @@
+     if opts.MonomialAlg then( 
+	 --for monomial CI it is enough to check if the sum of the variable is a Lefschetz element
+	 B:= R/ideal(I,sum for i to n-1 list L#i);
+	 return degree B == Sperner)
+     else(
+	 indx:=0;
+	 while HF#indx<=HF#(indx+1) do indx=indx+1;--increasing part of the Hilbert function
+	 while indx<e do if HF#indx<HF#(indx+1) then return false else indx=indx+1;--return false if the Hilbert function is not unimodal 
+    	 a:= symbol a;
+    	 T:=A[a_1..a_n];
+    	 X:= for i from 0 to n-1 list sub(L#(i-1),T);
+    	 line:= sum (for i from 1 to n list a_i*X#(i-1));
+    	 m:=sub(ideal(for i from 1 to n list X#(i-1)),T);
+    	 if opts.GorensteinAlg then (
+	     --for a Gorenstein, it is enough to check that there exists a linear l form such that the multiplication has max rank in the middle degree xl:A_(floor(e/2)->A_(floor(e/2)+1)
+	     lineScaled:= mingens m^(floor(e/2))*line;
+	     VarsScaled:= mingens m^(floor(e/2)+1);
+	     M:=lineScaled//VarsScaled; 
+	     d:=min(HF#(floor(e/2)),HF#(floor(e/2)+1));
+ 	     return rank M>= d;)
+     	 else (
+	      degreeSperner:=0;
+	      while HF#degreeSperner<Sperner do degreeSperner = degreeSperner+1;--degree for which the Hilbert function reaches maximum value
+	      mA:=ideal(for i from 0 to n-1 list sub(L#i,A));--the maximal homogeneous ideal of A
+	      socleA:= ann mA;
+	      indexAnn:=0;
+	      while  (HF#indexAnn-hilbertFunction(indexAnn, socleA))==0 do indexAnn=indexAnn+1;--annindex is the first degree where SocleA is non-zero, annindex<=e (for A Gorenstein they are equal)
+	      l:=min((indexAnn-1), degreeSperner); -- multiplication index
+    	      j:=0;-- flag
+     	      while (j<1 and l<degreeSperner+1) do (
+		  lineScaled1:= mingens m^(l)*line;
+	 	  VarsScaled1:= mingens m^(l+1);
+	 	  M1:=lineScaled1//VarsScaled1;
+	 	  d1:=min(HF#l,HF#(l+1));
+	 	  if  rank M1<d1 then j=1 else l=l+1); 
+     	     return j==0;)))
+
+
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+-- auxiliary function
+-- Input: integers n,s 
+-- output:list of n-tuples (a_1,....,a_n) such that a_1+.......+a_n-n=s and 2<=a_1<= ....<= a_n
+   
+   nTuples = (n, s) -> (
+    tuples:= {}; -- initialize an empty list to store the tuples
+    -- base case
+    if n == 1 then (
+        tuples = append(tuples, {s+1});
+        return tuples;)
+    else( -- recursive case
+    	for i from 2 to s+n//n do ( -- loop over possible values of the first component (from 1 to s)
+            subTuples := select(nTuples(n - 1, s - i+1), x -> x#0>=i); -- recursively generate (n-1)-tuples that sum to (s-i +n) and that the fist component is not less then i
+            iTuples := for j from 0 to #subTuples -1 list  {i}| subTuples#j;-- append i as the first component to each sub-tuple
+            tuples = tuples | iTuples;
```

Comment:
In general, repeatedly appending to a list inside a loop is a bad idea, since it generates a brand new list at every iteration.  Is it possible to re-factor this using `apply`,`for ... list`, or maybe modifying a mutable list in-place?

Reply:
This is still an issue -- the problem is that we keep re-creating `tuples` by appending to it with `|`.  Here's an alternate implementation using `apply`:

```m2
nTuples = (n, s) -> (
    if n == 1 then {{s + 1}}
    else flatten apply(toList(2..(s+n)//n), i -> (
	    subTuples := select(nTuples(n - 1, s - i+1), x -> x#0>=i);
	    apply(subTuples, x -> {i}|x))))
```